tests/src/mender-update/update_module/v3/update_module_test.cpp

```diff
@@ -216,6 +217,74 @@ class UpdateModuleTestWithDefaultArtifact {
 	unique_ptr<update_module::UpdateModule> update_module;
 };
 
+class UpdateModuleTestWithArtifactContainingIllegalPayloadFile {
+public:
+	UpdateModuleTestWithArtifactContainingIllegalPayloadFile(
+		UpdateModuleTests &tests, size_t mb = 1, size_t number_of_files = 1) {
+		// ASSERT doesn't work well inside constructors because of some peculiar return
+		// semantics, so wrap it in a lambda.
+		[&]() {
+			auto maybe_artifact = tests.PrepareArtifact(mb, number_of_files);
+			ASSERT_TRUE(maybe_artifact) << maybe_artifact.error();
+			auto artifact_file = maybe_artifact.value();
+
+			// Modify the only payload's file to point to an illegal path
+			std::filesystem::current_path(tests.temp_dir_.Path());
+			int result = std::system(("tar -xf " + artifact_file).c_str());
+			ASSERT_EQ(result, 0) << "Setup failed: Could not extract default artifact.";
+
+			result = std::system("tar -xf data/0000.tar.gz");
+			ASSERT_EQ(result, 0) << "Setup failed: Could not extract default artifact's payload.";
+
+			result = std::system("tar -zcf 0000.tar.gz --transform='s|rootfs|../rootfs|' rootfs");
+			ASSERT_EQ(result, 0)
+				<< "Setup failed: Could not recompress modified artifact's payload.";
+
+			result = std::system("mv 0000.tar.gz data/");
+			ASSERT_EQ(result, 0) << "Setup failed: Could not move modified artifact's payload.";
```

Comment:
Nit pick: this could be one single command:
```suggestion
			result = std::system("tar -zcf data/0000.tar.gz --transform='s|rootfs|../rootfs|' rootfs");
			ASSERT_EQ(result, 0)
				<< "Setup failed: Could not recompress modified artifact's payload.";
```

Reply:
Good point. I was trying to avoid calling tar with the file to be compressed in a different directory (e.g. data/rootfs which results in a "data/rootfs" inside the tarball), but this is different, here we have the resulting archive in a different directory.

So in fact this can be one command, thanks. Changed.